File: plot/src/main/java/net/fwitz/math/plot/color/complex/domain/ComplexColorWheel.java
package net.fwitz.math.plot.color.complex.domain;

import net.fwitz.math.complex.Complex;
import net.fwitz.math.plot.color.complex.ComplexColorFunction;

import java.awt.*;

import static java.awt.Color.getHSBColor;
import static java.lang.Math.floor;
import static java.lang.Math.log;
import static java.lang.Math.pow;

public class ComplexColorWheel implements ComplexColorFunction {
    private static final double TWO_PI = Math.PI * 2.0;

    private static float hue(Complex z) {
        double hue = z.arg() / TWO_PI;
        return (float) (hue - floor(hue));
    }

    private static double logScaleMantissa(Complex z) {
        double m = log(z.abs());
        return m - floor(m);
    }

    private static float saturation(double k) {
        double sat = (k < 0.5) ? k * 2 : 1 - (k - 0.5) * 2;
        sat = 1 - pow(1 - sat, 3);
        sat = 0.4 + sat * 0.6;
        return (float) sat;
    }

    private static float brightness(double k) {
        double val = k < 0.5 ? k * 2 : 1 - (k - 0.5) * 2;
        val = 1 - val;
        val = 1 - pow((1 - val), 3);
        val = 0.6 + val * 0.4;
        return (float) val;
    }

    public Color apply(Complex c, Complex z) {
        double k = logScaleMantissa(z);
        float hue = hue(z);
        float sat = saturation(k);
        float val = brightness(k);
        return getHSBColor(hue, sat, val);
    }
}

File: plot/src/main/kotlin/net/fwitz/math/plot/binary/complex/color/ComplexColorWheel.kt
package net.fwitz.math.plot.binary.complex.color

import net.fwitz.math.binary.complex.Complex
import java.awt.Color
import kotlin.math.floor
import kotlin.math.ln
import kotlin.math.pow

object ComplexColorWheel : ComplexColorFunction {
    private const val TWO_PI = Math.PI * 2.0

    private fun hue(z: Complex): Float {
        val hue: Double = z.arg / TWO_PI
        return (hue - floor(hue)).toFloat()
    }

    private fun logScaleMantissa(z: Complex): Double {
        val m = ln(z.abs)
        return m - floor(m)
    }

    private fun saturation(k: Double): Float {
        var sat = if (k < 0.5) k * 2 else 1 - (k - 0.5) * 2
        sat = 1 - (1 - sat).pow(3.0)
        sat = 0.4 + sat * 0.6
        return sat.toFloat()
    }

    private fun brightness(k: Double): Float {
        var brt = if (k < 0.5) k * 2 else 1 - (k - 0.5) * 2
        brt = 1 - brt
        brt = 1 - (1 - brt).pow(3.0)
        brt = 0.6 + brt * 0.4
        return brt.toFloat()
    }

    override fun invoke(c: Complex, z: Complex): Color {
        val k = logScaleMantissa(z)
        val hue = hue(z)
        val sat = saturation(k)
        val brt = brightness(k)
        return Color.getHSBColor(hue, sat, brt)
    }
}